In process test support

rbtcollins opened the pull request:
This is incremental work towards in-process tests.

rbtcollins commented:
so, here's the basic idea
read CONTRIBUTING
read the two new traits that are in their own files
the vast bulk is basically just adding <V: VarSource> .. (vars:&V,
to signatures
or to structs
finally changed DownloadCfg  to be Clone based rather than Copy.

I'm not 100% sure that the way I've done this is right.

For instance, we could have just one object that implements multiple traits; and we could have multiple traits combined into one aggregate to make it less noisy in code...

e.g. ProcessInterface: ArgsSource+VarSource + (process:&P...

which we could then break apart for when a narrow interface really is appropriate.

I'm happy enough to do a refactor like that if you like the sounds of it.

kinnison commented on download/src/lib.rs:
```diff
@@ -7,7 +7,7 @@ use url::Url;
 pub use crate::errors::*;
 
-#[derive(Debug, Copy, Clone)]
+#[derive(Debug, Clone)]
```
I'm not sure why you relaxed `Copy` here given it's a trivial enumeration.

rbtcollins replied:
Over-keen I think, I was trying to provoke closer-to-source errors on the trait. Happy to back it out; OTOH we don't need it anywhere either now, if we ever did.

kinnison commented on src/cli/argsource.rs:
```diff
@@ -0,0 +1,54 @@
+use std::env;
+
+pub trait ArgSource {
+    type Args: Iterator<Item = String>;
```
Do we really treat environment variables as always being `String` rather than `OsString`?

rbtcollins replied:
hmm, I missed an args_os call in proxy-mode

rbtcollins replied:
fixed

kinnison commented on src/cli/argsource.rs:
```diff
@@ -0,0 +1,54 @@
+    i: usize,
+}
+
+impl From<&Vec<String>> for VecArgs {
```
Would it be okay to have this as `From<&[String]>` and to have a complementary `From<&[&str]>` perhaps?  I tried to persuade an `impl<S: AsRef<str>> From<&[S]> for Foo` to exist, but it started whining that `&[&str; 2]` wasn't compatible which made me a bit sad.

rbtcollins replied:
.collect() on those should do the right thing for us already; this type is *much* easier as an owned type. I mean yes - but in my conversions in run() in the -wip branch, it isn't an issue.

kinnison commented on src/cli/argsource.rs:
```diff
@@ -0,0 +1,54 @@
+    }
+}
+
+impl Iterator for VecArgs {
```
I'd have been tempted to just set the iterator type for VecArgs to `Box<dyn Iterator<Item=String>>` and then return `Box::new(self.v.iter().cloned())` fom the `fn args()` but this is okay.

rbtcollins replied:
I tried things in that space. std::slice::Iterator<String> returns &String. Hmm, I can go back and noodle on that, or we can do as another pass to make it tighter. I think the contract is the key thing.

kinnison commented on src/cli/argsource.rs:
```diff
@@ -0,0 +1,54 @@
+    type Item = String;
+    fn next(&mut self) -> Option<String> {
+        if self.i == self.v.len() {
+            return None;
```
If you want to retain this implementation then it'd be right to also `impl FusedIterator` for this because you are fused.  Not that we rely on this but I think it does allow theoretically some optimisation.

rbtcollins replied:
it allows optimising .fuse(), so sure, we can; not an issue in practice I think

kinnison commented on src/cli/common.rs:
```diff
@@ -147,7 +148,10 @@ impl NotifyOnConsole {
-pub fn set_globals(verbose: bool, quiet: bool) -> Result<Cfg> {
+pub fn set_globals<V>(vars: &V, verbose: bool, quiet: bool) -> Result<Cfg<V>>
+where
+    V: VarSource + Clone,
```
Given how often you say `VarSource + Clone` would it not be easier to just supertrait it?  If we ever need to make a non-Clone VarSource we could revisit it.

rbtcollins replied:
By this point I was doing large scale grep-replace and not thinking :) - I only found out about the Clone need when I had the entire patch in place and was substituting in the HashMap layer in the WIP branch and found it was missing. Sure. I think perhaps the refactoring I refer to in the conversation is perhaps a bigger better rearrangement.

kinnison commented on src/cli/common.rs:
```diff
@@ -157,24 +161,27 @@ pub fn set_globals(verbose: bool, quiet: bool) -> Result<Cfg> {
 
-pub fn show_channel_update(
-    cfg: &Cfg,
+pub fn show_channel_update<V: VarSource + Clone>(
```
We could get rid of a lot of this noise by boxing the VarSource in the `Cfg` instance rather than parameterising it on it.

kinnison replied:
Also, thinking about it, that'd be faster to build too, and not be enough of a performance issue to be worth considering against IMO.

kinnison replied:
Then when you need to pass the var source to other things, make it take a `&dyn VarSource` instead?

rbtcollins replied:
thats a possibility too, yes.

kinnison commented on src/varsource.rs:
```diff
@@ -0,0 +1,38 @@
+use std::env;
+use std::ffi::{OsStr, OsString};
+
+pub trait VarSource {
```
I'm half tempted to suggest we do "typesafe environment variables" and gather all environment variable retrieval into one place, by putting a trait `RustupVars` which is implemented for any T which implements `VarSource` and then pass `&dyn RustupVars` around rather than `VarSource` -- perhaps that's a refactor for another time though.

rbtcollins replied:
not sure how that looks different

kinnison replied:
The idea would be that anywhere using `RustupVars` would call things like `.ram_limit()` and such, but the more I think about it the less I think it's needed right now since we don't tend to read the same variable from more than one place.

kinnison reviewed (commented):
Initial thoughts - I wouldn't **block** on them, but if you think there's merit then it might reduce the size of the diff somewhat.

rbtcollins committed:
Abstract over the current process

In order to do in-process testing process state - the current pid, args,
environment variables, cwd, and well known file descriptors and process
exit need to be abstracted out.

This patch introduces a system for doing so that is low (but not zero)
overhead, while relatively unintrusive.

A side effect of not calling exit() from deep in the process was to fix
a bug where downloads and tmp directory cleanup was not happening when
self updates were downloaded.

kinnison commented on src/bin/rustup-init.rs:
```diff
@@ -25,41 +24,46 @@ use rustup::cli::proxy_mode;
-        common::report_error(e);
-        std::process::exit(1);
-    }
+    let process = OSProcess {};
```
I'd prefer to see this implement `Default` or else have a trivial `new` method, just because this looks ugly.

rbtcollins replied:
sure, though for something used just once - shrug :). I will add.

kinnison commented on src/bin/rustup-init.rs:
```diff
@@ -25,41 +24,46 @@ use rustup::cli::proxy_mode;
+            common::report_error(e);
+            std::process::exit(1);
+        }
+        Ok(utils::ExitCode(c)) => std::process::exit(c),
```
Should this be `std::process::exit()` or should it really be `process.exit()` with that abstracted through the trait?

rbtcollins replied:
Because exit's occur deep in call stacks, process() can't have an exit() method unless we implement that as panic() - which is ugly, or a macro equivalent try! - which isn't a method - or we get into compiler plugins - which is opaque to readers.

We could downcast errors to extract ExitCode's from errors that have them, but errors can still be not propogated, and downcasting with error-chain is going to just make the migration from that more complicated I feel.

We could treat all errors as 'bad thing exit 1' and Ok(explicit error code) as a form of success, even if that means exiting non-zero e.g. for the proxy mode. Which is what I chose

Thus exit() is implemented as structured return: Err(badthing) for error class returns, and Ok(ExitCode(n)) for deliberate non-error returns where that is needed, which is here.

kinnison replied:
Fair enough.  I suppose we just don't do in-process testing of this part of `rustup-init` and we'll be fine. :+1:

rbtcollins replied:
Well we can't - its not in the library (it is in bin/ :) ). We could move some of this into the lib, but actually I think this is sufficiently shallow that bugs are remote.

kinnison commented on src/currentprocess.rs:
```diff
@@ -0,0 +1,228 @@
+}
+
+lazy_static! {
+    static ref HOOK_INSTALLED: Mutex<bool> = Mutex::default();
```
`Mutex<bool>` can often be replaced with an `AtomicBool` instance -- does this need the `Mutex` behaviour?

rbtcollins replied:
AtomicBool would permit a data race. set_hook is not a CAS operation, and orig_hook is global state.

kinnison commented on src/currentprocess.rs:
```diff
@@ -0,0 +1,228 @@
+            }));
+            *installed = true
+        }
+    }
```
This whole thing could be a `Once` perhaps?

kinnison reviewed (commented):
Looks fairly good, a few comments to deal with though.
